# Distance Contraction — Ion-Migration Pathway Screener

**Author:** Ramanuja Srinivasan Saravanan (University of Maryland)  
**Supervisor:** Prof. Yifei Mo  
**Version:** 1.0  
**License:** MIT  

---

## Overview

This notebook screens candidate ion-migration pathways in solid-state electrolyte frameworks
using the *distance-contraction* descriptor.

**The core idea in plain language:**  
When a mobile ion (Na⁺, Li⁺, K⁺, Mg²⁺, …) hops between two lattice sites, it
has to squeeze past neighbouring cations and anions. At the tightest point —
the transition state (TS) — those neighbours are *closer* than their equilibrium
bond lengths. We quantify that squeeze as a **distance contraction**: how much
shorter each bond is at the TS compared with a reference distance in the pristine
structure. A large contraction means a high energy barrier; a small contraction
means the path is relatively open and the ion can hop easily.

**Technically:**  
All mobile-ion site pairs within a user-defined hop-distance cutoff are first
enumerated. Crystal symmetry is then used to reduce these to a set of
*symmetry-unique* paths — pairs that are crystallographically inequivalent and
therefore have distinct energy barriers. The computed descriptor values are
subsequently mapped back onto *all* original pairs via the symmetry mapping,
so that the full percolation network (which requires every path, not just the
unique ones) can be reconstructed correctly for dimensionality analysis.

For each symmetry-unique path the notebook:
1. Generates linearly interpolated NEB images between the start and end sites.
2. Displaces the migrating ion transversely to minimise worst-case steric clash
   with neighbouring ions, producing a more physically realistic pathway geometry.
3. Evaluates four transverse configurations (cation-only neighbours vs. all
   neighbours, positive vs. negative direction) and selects the one with the
   smallest transition-state contraction.
4. Identifies the transition-state image as the one exhibiting maximum
   distance contraction relative to a reference (mainly *path*, and optionally *struct*, *stddev*).
5. Maps descriptor values back to all symmetry-equivalent paths and
   incrementally adds paths — sorted by contraction — to determine the
   dimensionality (1-D / 2-D / 3-D) at which the percolation network first
   becomes connected.

All data-loading, geometry, and pipeline code lives in **`dc_core.py`**
(imported in Section 1) — this notebook only holds setup, the run settings,
and the results display.

---

## How to use

There is nothing to click except **Run All** (Runtime/Kernel → Run all cells).
Every setting has a sensible default, printed to the output as it runs, so a
completely unmodified run analyzes the bundled example structure
(Li3YCl6 (LYC), a well-known Li superionic conductor) with
the mobile ion set to Li.

1. **Section 1** installs packages, downloads `dc_core.py` + the reference
   data files (periodic table, bond-distance statistics, and a handful of
   example CIFs — including the default one below), and imports `dc_core`.
2. **Section 2** is a single cell: a plain block of variables at the top —
   the mobile-ion element, the structure to analyze, and every other tunable
   parameter, each documented with what it does and its default value —
   followed by the pipeline run and all three result tables. Edit any
   variable, or leave them as-is, then run that one cell: it prints the
   settings it used, runs the analysis, and prints every result table at the
   end of its own output. There is nothing else to run afterward — it is one
   cell, top to bottom, every time you run it.

To analyze your **own** structure instead of the bundled example: in the
left sidebar, click the 📁 **Files** icon → ⬆ **Upload**, and add your CIF
file named `<ICSD_number>.cif` (e.g. `29962.cif`), then set `ICSD_NUMBER` in
Section 2 to that number. All other reference data is downloaded
automatically in Section 1 — no other uploads needed.

| File | Description |
|---|---|
| `<ICSD>.cif` or `<ICSD>_ord_rama.cif` | Crystal structure (CIF format) |
| `periodic_table.json` | Shannon ionic radii & full periodic table data |
| `Na_bond_stddev.csv`, `Li_bond_stddev.csv`, ... | Per-element bond-distance statistics |


## Section 1 — Environment Setup

Install packages and load the core library (`dc_core.py`).  
Run this cell once at the start of each Colab/Jupyter session.


In [ ]:
# ── Package installation (Colab) ──────────────────────────────────────────
# Skip if packages are already installed in your local environment.
!pip install --upgrade --quiet ase
!pip install --quiet pymatgen==2025.6.14 natsort


In [ ]:
# ── Download dc_core.py + reference data, then import ──────────────────────
import os, sys, subprocess
import pandas as pd                    # used directly by the Section 3 result-display cells
from IPython.display import display    # used directly by the Section 3 result-display cells

os.chdir("/content")

_MODULE_URL = "https://raw.githubusercontent.com/Rama-MSE-UMD/Distance-contraction-descriptor/main/dc_core.py"

if not os.path.exists("dc_core.py"):
    subprocess.run(["wget", "-q", _MODULE_URL])
if not os.path.exists("dc_core.py"):
    raise FileNotFoundError(
        "Could not fetch dc_core.py automatically.\n"
        "Either add it to the GitHub repo at:\n  " + _MODULE_URL + "\n"
        "or upload dc_core.py manually to /content (Colab: Files pane → upload), then re-run this cell."
    )

sys.path.insert(0, "/content")
import dc_core as dc

dc.download_base_files()
dc.load_reference_data()
print("dc_core loaded.")


⏭ Already exists: periodic_table.json
⏭ Already exists: Na_bond_stddev.csv
⏭ Already exists: Li_bond_stddev.csv
⏭ Already exists: K_bond_stddev.csv
⏭ Already exists: Mg_bond_stddev.csv
⏭ Already exists: Al_bond_stddev.csv
⏭ Already exists: Ca_bond_stddev.csv
⏭ Already exists: 39248_ord_rama.cif
⏭ Already exists: 1529809.cif
⏭ Already exists: 29962_ord_rama.cif
⏭ Already exists: 230141.cif
⏭ Already exists: 18640.cif
⏭ Already exists: 942733.cif

Done.
Reference databases loaded.
dc_core loaded.


## Section 2 — Settings, Run, and Results

The cells hold the settings, the pipeline run,
*and* the three result tables (the 3×6 descriptor summary, the unique-path
list, and the per-path overview) — running it does the whole analysis and
prints every result at the end of its own output. Leaving everything at its default analyzes the
bundled example structure (LYC, ICSD 29962) with the mobile ion set to
**Li** — you'll see that confirmed in the printed output.

| Variable | Meaning | Default |
|---|---|---|
| `MOBILE_ION` | Mobile-ion element symbol | `"Li"` |
| `ICSD_NUMBER` | ICSD code of the structure to analyze — expects `<ICSD>_ord_rama.cif` or `<ICSD>.cif` in the working directory | `"942733"` (bundled example) |
| `MAX_PATH_LENGTH` | Maximum hop distance considered, in Å | `7` |
| `NIMAGES` | Number of NEB interpolation images between two sites (excluding end-points) | `5` |
| `CREATE_FOLDER` | `"Yes"` writes every per-path CIF file into `<ICSD>_neb/`; `"No"` only writes the summary CSVs to the working directory | `"Yes"` |
| `COMPUTE_EXTRA` | `False` computes only the `path` reference category (faster); `True` also computes `struct` and `stddev` | `False` |

**Output files** (the run also prints a line for each one as it is written,
so you always know what landed where):

| File | Contents |
|---|---|
| `<ICSD>_data.csv` | Per-path distance-contraction metrics |
| `<ICSD>_deviation.csv` / `<ICSD>_dev.csv` | Per-structure dimensionality + contraction summary |
| `<ICSD>_uniquepair.csv` | All pairs with symmetry-mapped descriptor values |
| `<ICSD>_uniquesymm.csv` | Symmetry-unique paths only |
| `neb_<N>_linear.cif` | Linear (unoptimised) NEB path CIF |
| `neb_<N>_predicted_<mode>_<dir>.cif` | Transverse-optimised path CIFs |
| `image_0<N>.cif` | Individual NEB image structures |
| `neb_<N>_merged.cif` | All images of path *N* merged into one structure |
| `*Dim_struct_<label>.cif` | Mobile-ion network at each dimensionality threshold |

**Results shown at the end of the output:**

- **Table A**: 3 × 6 summary — contraction threshold (Å) at which each percolation dimensionality is first reached, for each reference scheme
- **Table B**: Unique NEB folders, clustered by contraction-value similarity
- **Table C**: Per-path contraction overview table (all paths, mapped via symmetry)


In [ ]:
# ── Section 2 settings — edit any value below, then run this cell ─────────
# See the table in the markdown cell above for what each one does.

MOBILE_ION      = "Li"       # mobile-ion element; one of: Li, Na, Mg, K, Ca, Al
ICSD_NUMBER     = "29962"    # structure to analyze (default = bundled example, Li3YCl6 / LYC)
MAX_PATH_LENGTH = 7          # Å, maximum hop distance considered
NIMAGES         = 5          # NEB interpolation images (excluding end-points)
CREATE_FOLDER   = "Yes"      # "Yes" -> also write per-path CIF files to <ICSD_NUMBER>_neb/
COMPUTE_EXTRA   = False      # True -> also compute the 'struct' and 'stddev' reference categories

# ════════════════════════════════════════════════════════════════════════════
# Everything below runs in this same cell — settings, the pipeline run, and
# all three result tables. One cell, one run, nothing else to execute.
# ════════════════════════════════════════════════════════════════════════════
element_oxi = dc.OXI_STATE.get(MOBILE_ION)
if element_oxi is None:
    raise ValueError(
        f"No oxidation state configured for '{MOBILE_ION}'. "
        f"Choose one of: {list(dc.OXI_STATE)}"
    )

print("Settings")
print(f"  Mobile ion       : {MOBILE_ION}  "
      f"(default is Li — edit MOBILE_ION above to change, e.g. 'Na', 'Mg', 'K')")
print(f"  Oxidation state  : {element_oxi}")
print(f"  ICSD structure   : {ICSD_NUMBER}")
print(f"  Max path length  : {MAX_PATH_LENGTH} Å")
print(f"  NEB images       : {NIMAGES}")
print(f"  Create folder    : {CREATE_FOLDER}")
print(f"  Categories       : {'Struct + stddev + path' if COMPUTE_EXTRA else 'Path only'}\n")

df_out, low_energy_df, df1, dim_struct, dim_path, dim_stddev, st_name, structure = dc.run_pipeline(
    element_symbol=MOBILE_ION,
    element_oxi=element_oxi,
    st_name=ICSD_NUMBER,
    max_path_length=MAX_PATH_LENGTH,
    nimages=NIMAGES,
    create_folder=CREATE_FOLDER,
    compute_extra=COMPUTE_EXTRA,
)




Settings
  Mobile ion       : Li  (default is Li — edit MOBILE_ION above to change, e.g. 'Na', 'Mg', 'K')
  Oxidation state  : 1
  ICSD structure   : 29962
  Max path length  : 7 Å
  NEB images       : 5
  Create folder    : Yes
  Categories       : Path only

Loaded: 29962_ord_rama.cif
Framework composition: Y6 Cl36
Total candidate pairs: 219
Symmetry-unique paths: 101 (reduced from 219)
Oxidation states: {'Li': 1.0, 'Y': 3.0, 'Cl': -1.0}
Primary anion: Cl (-1)
Mobile-ion → cation min distances: {'Li': 3.0005662455756728, 'Y': 3.765950893982475}

────────────────────────────────────────────────────────────
Path 1 / 101
  Path avg distances: {'Cl': 2.593666666666667, 'Li': 3.001, 'Y': 3.766}
  Best config: cation_pos  |  TS contraction = 0.820 Å
  [path]  max contraction = 0.820 Å  → High barrier
──────────────────────────────────────────────────

────────────────────────────────────────────────────────────
Path 2 / 101
  Path avg distances: {'Cl': 2.593666666666667, 'Li': 3.001, 'Y': 

,0-D,1-D,2-D,3-D,Struct,Composition
Reference,,,,,,
struct,NaN,NaN,None,NaN,29962,Li9 Y3 Cl18
path,0.44,0.44,None,0.63,29962,Li9 Y3 Cl18
stddev,NaN,NaN,None,NaN,29962,Li9 Y3 Cl18



9 unique paths out of 101 total (tol=0.02 Å):
['neb_27', 'neb_21', 'neb_18', 'neb_32', 'neb_92', 'neb_53', 'neb_30', 'neb_35', 'neb_50']



,NEB,NEB1,Category_path,Max_contraction_path
0,1,1,High,0.82
1,2,2,High,0.82
2,2,3,High,0.82
3,3,4,Low,0.61
4,4,5,High,0.79
5,5,6,Low,0.45
6,4,7,High,0.79
7,3,8,Low,0.61
8,5,9,Low,0.45
9,6,10,High,0.82


In [ ]:
# ── Results, Cell A — 3 × 6 descriptor summary table ───────────────────────
# Rows    : reference scheme (struct, path, stddev)
# Columns : dimensionality thresholds (0-D, 1-D, 2-D, 3-D), Struct, Composition
#
# Each value is the distance-contraction (Å) at which that dimensionality
# is first reached when paths are added in order of increasing contraction.
# None means that dimensionality was not achieved by any path combination.

def _get_dim(d, dim):
    val = d.get(dim, None)
    if val is None:
        return None
    try:
        return round(float(val), 2)
    except (TypeError, ValueError):
        return val

rows = []
for label, dim_dict in [
    ("struct", dim_struct),
    ("path",   dim_path),
    ("stddev", dim_stddev),
]:
    rows.append({
        "Reference":   label,
        "0-D":         _get_dim(dim_dict, 0),
        "1-D":         _get_dim(dim_dict, 1),
        "2-D":         _get_dim(dim_dict, 2),
        "3-D":         _get_dim(dim_dict, 3),
        "Struct":      st_name,
        "Composition": str(structure.composition),
    })

summary_df = pd.DataFrame(rows).set_index("Reference")
print("Distance contraction descriptor summary (Å):")
display(summary_df)

# ── Results, Cell B — unique NEB folders (clustered by contraction-value similarity) ──
unique_folders, cluster_map = dc.find_unique_neb_paths(
    df1, value_col="Max_contraction_path", tolerance=0.02)

print(f"\n{len(unique_folders)} unique paths out of {len(df1)} total (tol=0.02 Å):")
print(unique_folders)

# ── Results, Cell C — per-path contraction overview table (all paths, mapped via symmetry) ──
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

icsd_number = ICSD_NUMBER
data_csv = (f"{icsd_number}_neb/{icsd_number}_data.csv" if CREATE_FOLDER == "Yes"
            else f"{icsd_number}_data.csv")
df_results = pd.read_csv(data_csv)

numeric_cols = df_results.select_dtypes(include="number").columns
df_results[numeric_cols] = df_results[numeric_cols].round(2)

overview_cols = [
    "NEB", # "NEB1",
    "Category_stddev", "Max_contraction_stddev",
    "Category_path",   "Max_contraction_path",
]
print()
display(df_results[[c for c in overview_cols if c in df_results.columns]])

In [ ]:
# rm -r <ICSD>_neb   # uncomment + fill in to delete a run's output folder
